flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java

```diff
@@ -315,14 +326,26 @@ public FutureCompletingBlockingQueue<RecordsWithSplitIds<E>> getQueue() {
      * @throws Exception when failed to close the split fetcher manager.
      */
     public synchronized void close(long timeoutMs) throws Exception {
+        final long startTime = System.currentTimeMillis();
         closed = true;
         fetchers.values().forEach(SplitFetcher::shutdown);
+        // Actively drain the element queue in case there are previously shutting down
+        // fetcher threads blocking on putting batches into the element queue.
+        executors.submit(
+                () -> {
+                    while (fetchersToShutDown.get() > 0
+                            && System.currentTimeMillis() - startTime < timeoutMs) {
```

Comment:
Sorry for commenting after the patch is merged.

Better to use `System.nanoTime()` instead of `System.currentTimeMillis()` - Due to clock synchronization, it is possible that `System.currentTimeMillis() - startTime` be negative. Don't think this would cause any correctness or unexpected behavior issues. Just wanted to point that out.